Mathlib/Tactic/Linter/UnusedSetOptionIn.lean

```diff
@@ -0,0 +1,65 @@
+/-
+Copyright (c) 2024 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Tactic.Lemma
+import Lean.Elab.Command
+import Lean.Linter.Util
+
+/-!
+The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s.
+-/
+
+open Lean Elab Command
+
+/-- converts
+* `theorem x ...` to  `some (example ... , x)`,
+* `lemma x ...`   to  `some (example ... , x)`,
+* `example ...`   to ``some (example ... , `example)``,
+*  everything else goes to `none`.
+-/
+def toExample {m : Type → Type} [Monad m] [MonadRef m] [MonadQuotation m] :
+    Syntax → m (Option (Syntax × Syntax))
+  | `($dm:declModifiers theorem $did:declId $ds* : $t $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds* : $t $dv:declVal), did.raw[0])
+  | `($dm:declModifiers lemma $did:declId $ds* : $t $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds* : $t $dv:declVal), did.raw[0])
+  | `($dm:declModifiers example $ds:optDeclSig $dv:declVal) => do
+    return some (← `($dm:declModifiers example $ds $dv:declVal), mkIdent `example)
+  | _ => return none
+
+/-- The `unusedSetOptionIn` linter warns against "most external" unused `set_option ... in`s. -/
+register_option linter.unusedSetOptionIn : Bool := {
+  defValue := true
+  descr := "enable the unusedSetOptionIn linter"
+}
+
+/-- Gets the value of the `linter.unusedSetOptionIn` option. -/
+def getSetOptionIn (o : Options) : Bool := Linter.getLinterValue linter.unusedSetOptionIn o
+
+/-- reports a warning if the "first layer" `set_option ... in` is unused. -/
```

Comment:
How about this tweak?
"Report a warning if a `set_option ... in` command is unnecessary (i.e., the code elaborates fine without).
We only report the outermost `set_option ... in` (i.e., superfluous `set_option ... in`s which are nested are not linted against)."

If you'd like, you could also add
- why is not reporting on nested superfluous set_option in is not not so bad (is this just for simplicity, as the linter can be run multiple times? that's probably fine!)
- add one sentence why this is bad: for instance, might distract or confuse --- if this option was for backwards incompatibility, the fact that it can be removed now is a useful signal.

Reply:
The new doc-string is (once I will push)
```
Report a warning if a `set_option ... in` command is unnecessary
(i.e., the code elaborates fine without it).
We only report the outermost `set_option ... in` (i.e., nested, superfluous `set_option ... in`s
are not linted against).

This is useful since unnecessary `set_option ... in`s are often silent left-overs of adaptations
that are no longer needed.
Cleaning them up, helps maintaining healthy, up-to-date code.

The linter runs just on the outermost `set_option ... in` mostly for simplicity.
Consider whether emitting a warning if one of the nested `set_option ... in`s can be omitted,
by removing them one at a time.
```

Reply:
Looks great to me!